### PR TITLE
Monkeypatch flash attention in for llama

### PR DESCRIPTION
**Pull request by dakinggg**
This PR adds a monkeypatch of triton flash attention for llama2 models. If we start doing this for more models we can try to generalize into an algorithm, but until that time I think this monkeypatch is good enough.

TODO:
- [x] Copy profiling results for 7b
- [x] Copy profiling results for 70b
- [x] Paste in evidence that the test passes

Result for 7b scale to compare implementations:
<img width="433" alt="Screenshot 2023-08-10 at 4 31 57 PM" src="https://github.com/mosaicml/llm-foundry/assets/43149077/b453f0eb-42ae-4669-be52-3e5c72625f90">
<img width="456" alt="Screenshot 2023-08-10 at 4 32 04 PM" src="https://github.com/mosaicml/llm-foundry/assets/43149077/47f5aa13-b61e-48a4-a9ed-cc3833e55517">
<img width="448" alt="Screenshot 2023-08-10 at 4 32 20 PM" src="https://github.com/mosaicml/llm-foundry/assets/43149077/fb16ff2f-3420-445e-ab8f-b9a9166c5acf">

Test results:
```
tests/test_llama_patch.py::test_patch_equivalence[meta-llama/Llama-2-7b-hf-True-torch] 
------------------------------------------------------------------------------------------------------------------- live log call -------------------------------------------------------------------------------------------------------------------
2023-08-11 22:18:17 [    INFO] Setting seed to 42 (reproducibility.py:159)
2023-08-11 22:18:23 [    INFO] Setting seed to 42 (reproducibility.py:159)
PASSED                                                                                                                                                                                                                                        [ 12%]
tests/test_llama_patch.py::test_patch_equivalence[meta-llama/Llama-2-7b-hf-True-triton] 
------------------------------------------------------------------------------------------------------------------- live log call -------------------------------------------------------------------------------------------------------------------
2023-08-11 22:18:23 [    INFO] Setting seed to 42 (reproducibility.py:159)
2023-08-11 22:18:24 [    INFO] Setting seed to 42 (reproducibility.py:159)
PASSED                                                                                                                                                                                                                                        [ 25%]
tests/test_llama_patch.py::test_patch_equivalence[meta-llama/Llama-2-7b-hf-False-torch] 
------------------------------------------------------------------------------------------------------------------- live log call -------------------------------------------------------------------------------------------------------------------
2023-08-11 22:18:25 [    INFO] Setting seed to 42 (reproducibility.py:159)
2023-08-11 22:18:26 [    INFO] Setting seed to 42 (reproducibility.py:159)
PASSED                                                                                                                                                                                                                                        [ 37%]
tests/test_llama_patch.py::test_patch_equivalence[meta-llama/Llama-2-7b-hf-False-triton] 
------------------------------------------------------------------------------------------------------------------- live log call -------------------------------------------------------------------------------------------------------------------
2023-08-11 22:18:27 [    INFO] Setting seed to 42 (reproducibility.py:159)
2023-08-11 22:18:27 [    INFO] Setting seed to 42 (reproducibility.py:159)
PASSED                                                                                                                                                                                                                                        [ 50%]
tests/test_llama_patch.py::test_patch_equivalence[meta-llama/Llama-2-70b-hf-True-torch] 
------------------------------------------------------------------------------------------------------------------- live log call -------------------------------------------------------------------------------------------------------------------
2023-08-11 22:18:28 [    INFO] Setting seed to 42 (reproducibility.py:159)
2023-08-11 22:18:29 [    INFO] Setting seed to 42 (reproducibility.py:159)
PASSED                                                                                                                                                                                                                                        [ 62%]
tests/test_llama_patch.py::test_patch_equivalence[meta-llama/Llama-2-70b-hf-True-triton] 
------------------------------------------------------------------------------------------------------------------- live log call -------------------------------------------------------------------------------------------------------------------
2023-08-11 22:18:30 [    INFO] Setting seed to 42 (reproducibility.py:159)
2023-08-11 22:18:32 [    INFO] Setting seed to 42 (reproducibility.py:159)
PASSED                                                                                                                                                                                                                                        [ 75%]
tests/test_llama_patch.py::test_patch_equivalence[meta-llama/Llama-2-70b-hf-False-torch] 
------------------------------------------------------------------------------------------------------------------- live log call -------------------------------------------------------------------------------------------------------------------
2023-08-11 22:18:33 [    INFO] Setting seed to 42 (reproducibility.py:159)
2023-08-11 22:18:34 [    INFO] Setting seed to 42 (reproducibility.py:159)
PASSED                                                                                                                                                                                                                                        [ 87%]
tests/test_llama_patch.py::test_patch_equivalence[meta-llama/Llama-2-70b-hf-False-triton] 
------------------------------------------------------------------------------------------------------------------- live log call -------------------------------------------------------------------------------------------------------------------
2023-08-11 22:18:36 [    INFO] Setting seed to 42 (reproducibility.py:159)
2023-08-11 22:18:37 [    INFO] Setting seed to 42 (reproducibility.py:159)
PASSED                                                                                                                                                                                                                                        [100%]

================================================================================================================= warnings summary ==================================================================================================================
../../miniconda3/envs/foundry-3.10/lib/python3.10/site-packages/accelerate/utils/dataclasses.py:29
  /mnt/workdisk/danielking/miniconda3/envs/foundry-3.10/lib/python3.10/site-packages/accelerate/utils/dataclasses.py:29: DeprecationWarning: The distutils package is deprecated and slated for removal in Python 3.12. Use setuptools or check PEP 632 for potential alternatives
    from distutils.util import strtobool

../../miniconda3/envs/foundry-3.10/lib/python3.10/site-packages/deepspeed/ops/op_builder/builder.py:15
  /mnt/workdisk/danielking/miniconda3/envs/foundry-3.10/lib/python3.10/site-packages/deepspeed/ops/op_builder/builder.py:15: DeprecationWarning: The distutils.sysconfig module is deprecated, use sysconfig instead
    import distutils.sysconfig

../../miniconda3/envs/foundry-3.10/lib/python3.10/site-packages/jupyter_client/connect.py:20
  /mnt/workdisk/danielking/miniconda3/envs/foundry-3.10/lib/python3.10/site-packages/jupyter_client/connect.py:20: DeprecationWarning: Jupyter is migrating its paths to use standard platformdirs
  given by the platformdirs library.  To remove this warning and
  see the appropriate new directories, set the environment variable
  `JUPYTER_PLATFORM_DIRS=1` and then run `jupyter --paths`.
  The use of platformdirs will be the default in `jupyter_core` v6
    from jupyter_core.paths import jupyter_data_dir, jupyter_runtime_dir, secure_write

../../miniconda3/envs/foundry-3.10/lib/python3.10/site-packages/comet_ml/monkey_patching.py:19
  /mnt/workdisk/danielking/miniconda3/envs/foundry-3.10/lib/python3.10/site-packages/comet_ml/monkey_patching.py:19: DeprecationWarning: the imp module is deprecated in favour of importlib and slated for removal in Python 3.12; see the module's documentation for alternative uses
    import imp

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
========================================================================================================== 8 passed, 4 warnings in 22.00s ===========================================================================================================
```

70b running on 16 and 32 gpus:
<img width="899" alt="Screenshot 2023-08-11 at 2 59 36 PM" src="https://github.com/mosaicml/llm-foundry/assets/43149077/d6ee425d-4d88-44f9-80af-39a0e99934e6">
<img width="443" alt="Screenshot 2023-08-11 at 2 59 45 PM" src="https://github.com/mosaicml/llm-foundry/assets/43149077/26edfb35-e8a9-446c-9c48-528eac5571de">
<img width="464" alt="Screenshot 2023-08-11 at 2 59 50 PM" src="https://github.com/mosaicml/llm-foundry/assets/43149077/797f0319-57a2-4052-8bc0-a37accf2d1cd">
<img width="431" alt="Screenshot 2023-08-11 at 3 00 00 PM" src="https://github.com/mosaicml/llm-foundry/assets/43149077/2f82e12d-b38e-4831-8266-a880b092bcba">

### CODE REVIEW

File: tests/test_llama_patch.py [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+        # but the torch implementation produces the exact same output so we can confirm
+        # the implementation is correct
+        patch_fn = llama_attention_patch_triton
+        dtype = torch.bfloat16
+        atol = 1e-2
+        rtol = 1e-2
+    else:
+        raise ValueError(f'Unknown patch_fn_name: {patch_fn_name}')
+
+    llama_7b_config = transformers.AutoConfig.from_pretrained(
```
Comment (vchiley) [inline tests/test_llama_patch.py:51]:
```suggestion
    llama_config = transformers.AutoConfig.from_pretrained(
```

since `model_name` can be 70B as well.

File: tests/test_llama_patch.py [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+        model_name, use_auth_token=True)
+
+    reproducibility.seed_all(42)
+    attention = LlamaAttention(config=llama_7b_config,)
```
Comment (vchiley) [inline tests/test_llama_patch.py:55]:
```suggestion
    attention = LlamaAttention(config=llama_config,)
```

File: tests/test_llama_patch.py [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+        hidden_states=hidden_states,
+        attention_mask=causal_mask if explicit_mask else None,
+        position_ids=None,
+        past_key_value=None,
+        use_cache=False,
+    )
+
+    reproducibility.seed_all(42)
+    LlamaAttention.forward = patch_fn
+    attention = LlamaAttention(config=llama_7b_config,)
```
Comment (vchiley) [inline tests/test_llama_patch.py:82]:
```suggestion
    attention = LlamaAttention(config=llama_config,)
```

File: tests/test_llama_patch.py [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+    new_output, _, _ = attention(
+        hidden_states=hidden_states,
+        attention_mask=causal_mask if explicit_mask else None,
+        position_ids=None,
+        past_key_value=None,
+        use_cache=False,
+    )
+
+    # Reset the forward function so patches don't persist
+    LlamaAttention.forward = original_forward
```
Comment (vchiley) [inline tests/test_llama_patch.py:93]: Is this needed?
doesn't test end destruct `LlamaAttention`?

File: llmfoundry/models/layers/llama_attention_monkeypatch.py [hunk 5]
```diff
@@ -0,0 +1,283 @@
...
+        query_states, key_states, cos, sin,
+        position_ids)  # type: ignore (thirdParty)
+
+    ### MAIN MODIFICATIONS START HERE ###
+    query_states = query_states.transpose(1, 2).view(
+        bsz, q_len, self.num_heads * self.head_dim)
+    key_states = key_states.transpose(1, 2).view(
+        bsz, q_len, self.num_key_value_heads * self.head_dim)
+    value_states = value_states.transpose(1, 2).view(
+        bsz, q_len, self.num_key_value_heads * self.head_dim)
```
Comment (vchiley) [inline llmfoundry/models/layers/llama_attention_monkeypatch.py:136]: lines 115 thru 120 do the same transpose

is there anyway to not do it there so we dont have to undo it?
ie is there any way to apply RoPE along diff dim?

File: llmfoundry/models/layers/llama_attention_monkeypatch.py [hunk 6]
```diff
@@ -0,0 +1,283 @@
...
+        query=query_states,
+        key=key_states,
+        value=value_states,
+        n_heads=self.num_heads,
+        kv_n_heads=self.num_key_value_heads,
+        past_key_value=None,
+        softmax_scale=None,
+        attn_bias=attention_mask,
+        key_padding_mask=None,
+        is_causal=False,  # The causal mask is propagated from LLamaForCausalLM
```
Comment (vchiley) [inline llmfoundry/models/layers/llama_attention_monkeypatch.py:261]: if attention_mask is just the causal mask, its more performant to `del attention_mask`, set `attn_bias=None`, and set `is_causal=True`

File: llmfoundry/models/layers/llama_attention_monkeypatch.py [hunk 7]
```diff
@@ -0,0 +1,283 @@
...
+
+
+def get_llama_attention_patch_fn(patch_fn_name: str = 'torch') -> Callable:
+    if patch_fn_name == 'torch':
+        return llama_attention_patch_torch
+    elif patch_fn_name == 'triton':
+        return llama_attention_patch_triton
+    else:
+        raise ValueError(
+            f'Unrecognized llama attention patch function: {patch_fn_name}')
```
Comment (vchiley) [inline llmfoundry/models/layers/llama_attention_monkeypatch.py:64]: y not do a flash attn patch as well?

**Review by vchiley — COMMENTED** [review]
left a few comments

**Comment by germanjke** [comment]
Hi guys! @vchiley @dakinggg when do you planning to release this request? Thanks

File: llmfoundry/models/layers/llama_attention_monkeypatch.py (hunk 6 above)
Comment (dakinggg) [reply llmfoundry/models/layers/llama_attention_monkeypatch.py:261]: Agreed, but the `attention_mask` can be anything in huggingface, so isn't guaranteed to be just the causal mask, and doing it the way you describe could introduce some nasty silent errors.

File: tests/test_llama_patch.py (hunk 4 above)
Comment (dakinggg) [reply tests/test_llama_patch.py:93]: It is needed. The monkeypatch affects all uses within a process, and pytest runs in a single process.

File: llmfoundry/models/layers/llama_attention_monkeypatch.py (hunk 5 above)
Comment (dakinggg) [reply llmfoundry/models/layers/llama_attention_monkeypatch.py:136]: Yes, requires modifying a bit more code, and also monkeypatching/changing the Rope implementation in hf. My goal here was to minimally change the existing code to patch in flash attention. Do you think avoiding the extra transpose is compelling enough to pull in and modify more code?

**Review by vchiley — APPROVED** [review]
lgtm
